bombe/src/main/java/org/cadixdev/bombe/jar/Jars2.java

```diff
@@ -0,0 +1,148 @@
+
+/**
+ * Utilities for working with jar files.
+ *
+ * @author Jamie Mansfield
+ * @since 0.4.0
+ */
+public final class Jars2 {
+
+    public static Stream<AbstractJarEntry> walk(final FileSystem fs) throws IOException {
+        final Path root = fs.getPath("/");
+
+        final List<AbstractJarEntry> entries = new ArrayList<>();
+
+        Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
+            @Override
+            public FileVisitResult visitFile(final Path file, final BasicFileAttributes attrs) throws IOException {
+                entries.add(read(file));
+                return FileVisitResult.CONTINUE;
+            }
+        });
+
+        return entries.stream();
+    }
+
+    /**
+     * Creates an {@link AbstractJarEntry} for a given path.
+     *
+     * @param entry The path of the jar entry
+     * @return The jar entry
+     * @throws IOException Should the entry fail to read
+     */
+    public static AbstractJarEntry read(final Path entry) throws IOException {
+        try (final InputStream is = Files.newInputStream(entry)) {
+            final String name = entry.toString().substring(1); // Remove '/' prefix
+            final long time = Files.getLastModifiedTime(entry).toMillis();
+
+            if (Objects.equals("META-INF/MANIFEST.MF", name)) {
+                return new JarManifestEntry(time, new Manifest(is));
+            }
+            else if (name.startsWith("META-INF/services/")) {
+                final String serviceName = name.substring("META-INF/services/".length());
+
+                final ServiceProviderConfiguration config = new ServiceProviderConfiguration(serviceName);
+                config.read(is);
+                return new JarServiceProviderConfigurationEntry(time, config);
+            }
+
+            final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            ByteStreams.copy(is, baos);
+
+            if (name.endsWith(".class")) {
+                return new JarClassEntry(name, time, baos.toByteArray());
+            }
+            else {
+                return new JarResourceEntry(name, time, baos.toByteArray());
+            }
+        }
+    }
+
+    public static void transform(final FileSystem input, final Path output, final JarEntryTransformer... transformers) throws IOException {
+        final URI uri = URI.create("jar:" + output.toUri());
+        final Map<String, String> options = new HashMap<String, String>() {
+            {
+                this.put("create", "true");
+            }
+        };
+
+        try (final FileSystem fs = FileSystems.newFileSystem(uri, options)) {
+            for (final AbstractJarEntry entry : walk(input)
+                    .map(entry -> {
+                        for (final JarEntryTransformer transformer : transformers) {
+                            entry = entry.accept(transformer);
+                        }
+                        return entry;
+                    }).collect(Collectors.toList())) {
```

Comment:
I suspect this is acceptable because `.forEach` does not allow propagating the `IOException`s thrown below, but in general, collecting just to iterate over it is a bit meh.

Reply:
Again, I agree - those damn exceptions :(